### PR TITLE
Fix: typing notifications are not properly dismissed. 

### PR DESCRIPTION
**Pull request by jainkuniya**
Dismiss typing notification after 15s if no new event is received.

Fix:#1195

### CODE REVIEW

**Comment by vishwesh3** [comment]
@borisyankov  Updated 👍

**Comment by vishwesh3** [comment]
thanks @zulipbot Rebased 👍

File: src/config.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ export default {
   messagesPerRequest: 20,
   scrollCallbackThrottle: 500,
   messageListThreshold: 500,
+  typingNotificationTimeout: 15 * 1000,
```
Comment (borisyankov) [inline src/config.js]: There is no point in adding this to the config. We should hard code it.
Also, we likely want to remove more things from here later like messageListThreshold etc.

File: src/actionConstants.js [hunk 2]
```diff
@@ -83,3 +83,5 @@ export const DELETE_OUTBOX_MESSAGE = 'DELETE_OUTBOX_MESSAGE';
 
 export const DRAFT_ADD = 'DRAFT_ADD';
 export const DRAFT_REMOVE = 'DRAFT_REMOVE';
+
+export const LOCAL_EVENT_CLEAR_TYPING_NOTIFICATIONS = 'LOCAL_EVENT_CLEAR_TYPING_NOTIFICATIONS';
```
Comment (borisyankov) [inline src/actionConstants.js]: I don't think there is a need to add 'LOCAL_EVENT' to the name.

File: src/events/eventActions.js [hunk 3]
```diff
@@ -43,12 +44,16 @@ export const startEventPolling = (queueId: number, eventId: number) => async (
...
 
     const actions = [];
+    let startTypingNotificationLoop = false;
     for (const event of response.events) {
       lastEventId = Math.max(lastEventId, event.id);
       eventMiddleware(getState(), event);
       const action = eventToAction(getState(), event);
 
       if (action) {
+        startTypingNotificationLoop = !!(
```
Comment (borisyankov) [inline src/events/eventActions.js]: The `!!` is not necessary here. It is used to convert to 'boolean' and we already have that.

File: src/events/eventActions.js [hunk 4]
```diff
@@ -43,12 +44,16 @@ export const startEventPolling = (queueId: number, eventId: number) => async (
...
+          action.type === 'EVENT_TYPING_START' && Object.keys(getState().typing).length === 0
```
Comment (borisyankov) [inline src/events/eventActions.js]: Also, the place for this is in some action creator related to EVENT_TYPING_START. (which we currently do not have)

File: src/events/eventActions.js [hunk 5]
```diff
@@ -60,5 +65,10 @@ export const startEventPolling = (queueId: number, eventId: number) => async (
     } else if (actions.length === 1) {
       dispatch(actions[0]);
     }
+
+    if (startTypingNotificationLoop) {
+      // first typing event after a while so start clearTypingNotification loop
+      dispatch(clearTypingNotification());
```
Comment (borisyankov) [inline src/events/eventActions.js]: This whole file should be more generic and should not have knowledge about any specific event type.

File: src/typing/typingSelectors.js [hunk 6]
```diff
@@ -25,6 +25,6 @@ export const getCurrentTypingUsers = createSelector(
       return undefined;
     }
 
-    return currentTyping.map(userId => getUserById(users, userId));
+    return currentTyping.userIds.map(userId => getUserById(users, userId));
```
Comment (borisyankov) [inline src/typing/typingSelectors.js]: Hmmm, is this a bug fix?

File: src/typing/clearTypingNotification.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+/* @flow */
+import type { Dispatch, GetState } from '../types';
+
+import { getTyping } from '../directSelectors';
+import config from '../config';
+
+export const clearTypingNotification = () => async (dispatch: Dispatch, getState: GetState) => {
```
Comment (borisyankov) [inline src/typing/clearTypingNotification.js]: Hmm... this looks like quite a convoluted way to implement this.
Would you elaborate why a simpler `setTimeout` call is not enough?

File: src/typing/typingSelectors.js (hunk 6 above)
Comment (vishwesh3) [reply src/typing/typingSelectors.js]: Now we have `userIds` array and `time` at which event have arrived in `currentTyping`.

File: src/typing/clearTypingNotification.js (hunk 7 above)
Comment (vishwesh3) [reply src/typing/clearTypingNotification.js]: I found some difficulties with `timeout`. 
What we want to only dismiss those notifications which are outdated. For example if typing start event arrived at time `t` then it should only be dismissed at time `t+15 sec` if no new start typing event have arrived in between `t` and `t+15 sec`. If new start typing has arrived at `t+5 sec` then we need to take account at `t+5+15 sec` thus invalidating previous `timeout`.

Comment (vishwesh3) [reply src/typing/clearTypingNotification.js]: I found this to be more simpler than `setTimeout` one. :)

**Comment by vishwesh3** [comment]
rebases 👍

**Comment by vishwesh3** [comment]
@borisyankov thoughts on this one? :)

**Comment by jainkuniya** [comment]
@borisyankov rebased 👍 
Let me know changes :)

**Comment by borisyankov** [comment]
I think it looks good now. Good job 👍

**Pull request merged**